Madlibs

lz2003 opened the pull request:
Two versions of the game:
1. Default mad libs -> Set default story
2. Custom mad libs -> Upload your own story

I've played the custom game like twice cause making stories is hard, so tell me if anything unwanted issues happen. Used a little bit of colour. I also left some niche things for members to explore/extend their knowledge for the custom mad libs game, but I can add them in if wanted.

albertlai431 commented on madlibs/madlibs.py:
```diff
@@ -0,0 +1,69 @@
+  print("\n\nTHE END\n")
+
+# Template story using a list
+story_temp = ["One ", "ADJ", " Wednesday afternoon, ", "NAME", " decides to attend an exciting PETCS meeting! ", "NAME", " happily walks into the ", "ADJ", " classroom choosing to ", "VERB", " by their favourite computer. The execs warmly greet all attending members and announce the lesson they are teaching: ", "LESSON", "! Everyone is puzzled, but agree to ", "VERB", " to the lesson. Halfway thr ...
```
Should add a space between `"ADJ", "NOUN1"`

albertlai431 replied:
The verb after "the principal" should be singular

albertlai431 commented on madlibs/madlibs.py:
```diff
@@ -0,0 +1,69 @@
+    GREEN = '\033[92m'
+    END = '\033[0m'
+
+print("\nWelcome to PETCS Mad Libs!")
+
+keepPlaying = True
```
It would make more sense to put these lines on line 49/after all the functions are declared

albertlai431 commented on madlibs/madlibs.py:
```diff
@@ -0,0 +1,69 @@
+copy = story_temp[:]
+
+# Main game loop
+while keepPlaying == True:
```
No need for `== True`

lz2003 replied:
Thanks, completely missed that!

albertlai431 commented on madlibs/madlibs.py:
```diff
@@ -0,0 +1,69 @@
+# Main game loop
+while keepPlaying == True:
+  name = bcolours.GREEN + input("\nPlease name your character: ") + bcolours.END
+  print()
```
This could be removed and moved to line 22  --> `input("\nEnter an adjective: ")`

albertlai431 commented on madlibs/madlibs_custom.py:
```diff
@@ -0,0 +1,126 @@
+      # Hint: Try Except
+      if choice == "txt":
+        # Although we didn't teach open(), it is a built-in function for opening files
+        open_story = open(input("Enter file name: "), "r")
```
For me, this code errors if I just type in `example.txt`, I have to type in `madlibs/example.txt` - was this intentional? Since it sort of defeats the purpose of having it in the same directory.

lz2003 replied:
Weird, when I put example.txt it works? I think you have to change directories before running the script, so I'll probably just remove the option of opening a .txt file and have them copy/paste it in to avoid confusion.

albertlai431 replied:
oh rightt yes it's the directory I forgot oops! I was using repl.it for that and the directory is automatically the root directory so I think that's why it didn't work. But yeah, I think it might be better to just remove that to avoid confusion, though it was a good idea!

albertlai431 commented on madlibs/madlibs_custom.py:
```diff
@@ -0,0 +1,126 @@
+    GREEN = '\033[92m'
+    END = '\033[0m'
+
+print("\nWelcome to PETCS Mad Libs!\n")
+print('''Info --------------------->
+Copy/paste a custom 1 paragraph story that follows the legend!
+If you want to input a .txt file, make sure the .txt file is in the same directory (folder) as the python script!
+
+Recommended story length is 100 - 500 words, but I won't stop you in testing the limits :) 
+
+Legend for user input:
+ADJ - Ask for an adjective
+NOUN - Ask for a noun
+VERBP - Ask for a plural verb
+VERBS - Ask for a singular verb
+VERBD - Ask for a past tense verb
+ADVERB - Ask for a adverb
+* Legend is case sensitive!
+''')
+  
+keepPlaying = True
+repeat = False
+
+# Template story
+story_temp = []
+copy = []
```
Same comment as for `madlibs.py`, should be included after all the functions are declared

albertlai431 commented on madlibs/madlibs_custom.py:
```diff
@@ -0,0 +1,126 @@
+        story = open_story.read()
+        break
+      elif choice == "man":
+        story = input("Type/paste your story here: ")
```
Should probably make it clear that while typing the story then shouldn't use the `enter` key?

lz2003 replied:
Yeah good point. I'll specify that!

albertlai431 commented on madlibs/madlibs_custom.py:
```diff
@@ -0,0 +1,126 @@
+        return False
+
+# Sometimes an input prompt will be muddled with punctuation, so we should remove it!
+def remove_punct(word):
```
A couple of issues with the function:
- What if there are multiple types of punctuation used? The return statement will exit the function the first time it finds any type of punctuation
- The code assumes that the punctuation is at the end, but what if it was at the start?
- Opinion: I don't think it was necessary to use a set for `punct` LOL but isok 😛

IMO it would be easier to just do something like `if "ADJ" in story[i]` and then just replace the "ADJ" with the user input

lz2003 replied:
Yikes, completely forgot that 'if "ADJ" in story[i]' works, even with punctuation! Yep, going to definitely change that and get rid of remove_punct() function.

albertlai431 commented on madlibs/madlibs_custom.py:
```diff
@@ -0,0 +1,126 @@
+  return story.split()
+
+# Asks the user if they want to repeat the last story inputted
+def repeat_story():
```
Just to keep consistency, you should add an else statement to `print("Invalid input, try again!")`

albertlai431 replied:
Also idk if its necessary to create a separate function just for this but that's imo

lz2003 replied:
Yeah, I only made it a separate function because I don't like when you have too many nested loops together. TBH, you're 100% right that there is no use for the function, just me being stubborn, I'll remove it and just add it to the game loop.

lz2003 replied:
Actually, none of the functions are needed since they are only called once... I just think the functions separate important parts of the code, making it look cleaner for the user 😅

albertlai431 replied:
Yeah that's true LOL I think the way I see it is if the function should perform some sort of separate "unit" task separate from the game loop, but that's just an opinion haha

albertlai431 reviewed (changes requested):
**General Comments:**
- Overall solid work!
- Logic is sound and code is generally well commented
- The way you implemented the array stuff was pretty smart, and the open text file option was pretty cool too!
- No large bugs/errors with the code, but some small issues here and there

**To fix:**
- [ ] `madlibs.py`, `madlibs_custom.py`: inline comments (mostly just for the text, no logic/programmatic issues)
- By convention, functions are commented with [docstrings](https://stackoverflow.com/questions/2357230/what-is-the-proper-way-to-comment-functions-in-python)
- [ ] There are some discrepancies between the example.txt version and the version implemented in the `madlibs.py` - most notably the singular/plural of the verbs.

**Optional changes:**
- [ ] `madlibs.py`: I think it would sound smoother if you asked for the main character's gender at the beginning since it sounds a bit strange with the their/they. Maybe give the option for male, female, and unspecified? 
  - I guess this would be extra work for the `madlibs_custom.py` version... I think just implementing it only for the `madlibs.py` version would be okay. Unless you can think of a way to do it for the custom version :)
- [ ] I feel like the way you make copies of the `story_temp` array is too complicated, wouldn't it just be simpler to initialize a new copy with every iteration of the game loop rather than having to create a copy outside of the game loop and then having to recopy that copy? (note: I wrote this before reading the custom file, but I still think it could also be simplified as well)
- [ ] LMAO just an idea but would be cool to besides stuff like ADJ, VERB, etc make an option for emoji as well?? 😃

Edit: Ohh hmmmm looks like some of the in-line comments are down below as well 👇 oh well...

lz2003 commented:
### **Things I fixed/changed:**
- How the program replaces legend words (Adjective, Verbs, etc)
- Got rid of the redundant copy variable 
- Docstring comments in functions
- Opening .txt file was confusing, so I dropped it
- Added a confirm input story just in case the user accidentally presses enter
- Probably the most important fix: added EMOJI option to the legend
- I THINK I fixed the grammar issues in the example stories (This is why I get low marks in English LOL)
- Initially I was gonna do the gender thing, but honestly the story used they/their/them like twice, so it feels kind of pointless and the custom version does not have a NAME option in the legend anyways

albertlai431 reviewed (approved):
Everything looks good, all the issues were fixed properly! Even the *grammer* 😂

albertlai431 commented:
@lz2003 everything looks good and makes sense so I don't think Eddie will have to look over it, so you can merge the branch now!
